Feature 6620 implement td sequential candlestick indicator

rathnagiri opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

#### Description
This PR introduces the TdSequential technical indicator into the Lean framework.
It implements the TD Sequential algorithm with support for:

TD Buy/Sell Setup phases (up to 9 steps)

Perfect Setup detection

TD Buy/Sell Countdown phases (up to 13 steps)

Countdown cancellation via TDST resistance/support levels

State encoding of phase and step into a single decimal for easy usage in algorithms

The indicator is implemented as a subclass of IndicatorBase<TradeBar> and encodes the state using the format:
```
<phaseCode>.<step>
e.g., 1.04 = TD Buy Setup, Step 4
      5.09 = TD Buy Setup Perfect at Step 9
      6.13 = TD Buy Countdown Completed
```

#### Motivation and Context
This change adds support for the widely-used TD Sequential indicator, which helps traders identify trend exhaustion and potential reversal points.
The indicator is commonly used in professional trading environments and is now available to Lean algorithm authors with encoded values for seamless integration.

#### Requires Documentation Change
Yes, An update is required to include the TdSequential indicator in:
Algorithm.Indicators.cs for auto-initialization
Indicator documentation for usage patterns and interpretation of encoded values

#### How Has This Been Tested?
Tests verified using broad unit tests covering all scenario outputs related to each phase of the output

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
<!--- The following is a checklist of items that MUST be completed before a PR is accepted -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. <!--- If not applicable, please explain why -->
- [ ] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

Many of the tests fail out of the box without any of my changes.
<!--- Template inspired by https://www.talater.com/open-source-templates/#/page/99 -->

Martin-Molinero commented on Indicators/TdSequential.cs:
```diff
@@ -0,0 +1,313 @@
+        /// <inheritdoc />
+        public int WarmUpPeriod => 5;
+
+        /// <inheritdoc />
```
Sorry our documentation engine does not support `/// <inheritdoc />` atm, we just use the full version instead

rathnagiri replied:
done

Martin-Molinero commented on Indicators/TdSequential.cs:
```diff
@@ -0,0 +1,313 @@
+        private bool _inSellCountdown;
+
+        private decimal _tdstResistance; // higeest high of the 9-bar TD Sequential buy setup (indicates resistance)
+        private decimal _tdstSupport; // lowest low of the 9-bar TD Sequential sell setup (indicates support)
```
minor but `_tdstSupport/_tdstResistance` maybe we can rename these to `_supportPrice/_resistancePrice` or level if it's not a price

rathnagiri replied:
Thanks for the suggestion!.  I have renamed the fields to _resistancePrice and _supportPrice for clarity, as they store actual support or ressitance price levels from the 9-bar setups.

Martin-Molinero reviewed (changes requested):
Hi @rathnagiri! Thank you for the contribution 💪 

Leaving a few comments around, TLDR: I think there's some room for clean up and consolidation to make the indicator cleaner and clearer

rathnagiri commented on Indicators/TdSequential.cs:
```diff
@@ -0,0 +1,313 @@
+using QuantConnect.Data.Market;
```
done

rathnagiri commented on Indicators/TdSequential.cs:
```diff
@@ -0,0 +1,313 @@
+namespace QuantConnect.Indicators
+{    
+    /// <summary>
+    /// Represents the TD Sequential indicator, which is used to identify potential trend exhaustion points.
```
done

rathnagiri commented on Indicators/TdSequential.cs:
```diff
@@ -0,0 +1,313 @@
+    /// The TD Sequential indicator is a popular technical analysis tool that helps traders identify potential reversal points
+    /// in the market. It consists of two main components: the setup phase and the countdown phase. The setup phase counts the number of consecutive bars that meet specific criteria, while the countdown phase
+    /// counts the number of bars that follow the setup phase to confirm a potential reversal.
+    /// This implementation focuses on the setup phase, counting the number of consecutive bars that meet the criteria for a bullish or bearish setup.
```
nice catch., missed updating the header after my first implementation that focussed just on the setup algos

rathnagiri commented on Indicators/TdSequential.cs:
```diff
@@ -0,0 +1,313 @@
+        /// <summary>
+        /// Max Step count in Countdown phase
+        /// </summary>
+        public const int MaxCountdownCount = 13;
```
> Both these const seem to be a setting which should be passed as constructor argument, right? 9/13 could be the default values used at `Algorithm/QCAlgorithm.Indicators.cs`, can see other indicators as a pattern to follow Also don't think they should be public but just private readonly fields

@Martin-Molinero Thank you for the suggestion. You're right in pointing out that constants generally be private unless there’s a need for public access. I've made both MaxSetupCount and MaxCountdownCount private.

However, I've kept them as constants rather than making them configurable via the constructor. This is because the values 9 and 13 are fundamental to the Tom Demark Sequential algorithm as originally defined and are not intended to be adjusted. Making them configurable could introduce confusion or unintended behavior that deviates from the indicator's standard interpretation.

Let me know if you think it would help to still make this flexible in future use cases, but for now, keeping these values fixed seems the most correct and aligned with the intent of the algorithm.

rathnagiri commented on Indicators/TdSequential.cs:
```diff
@@ -0,0 +1,313 @@
+        /// <param name="input">The current trade bar input.</param>
+        /// <returns>The encoded state of the TD Sequential indicator for the current bar.</returns>
+        protected override decimal ComputeNextValue(TradeBar input)
+        {
+            _bars.Add(input);
+
+            if (_bars.Count < 5)
+                return Default;
+
+            if (_bars.Count > 10)
+                _bars.RemoveAt(0);
```
Thanks for the great suggestion @Martin-Molinero ! I've refactored the indicator to inherit from WindowIndicator<IBaseDataBar>, which might simplify the sliding-window logic and keeps the implementation more maintainable and I need to test, If not will use RollingWindow. Looks like WindowIndicator also integrates cleanly with IIndicatorWarmUpPeriodProvider and thereby removes the need to manually manage the bar buffer, aligning better with Lean's indicator design patterns. Also for style, added {} braces across single line control blocks .,

rathnagiri replied:
> These values seems a bit magical, what/why 10/5, same for WarmUpPeriod and IsReady, should this be an indicator setting too with a default value?

@Martin-Molinero 
I've replaced the hardcoded values (5 and 10) with named constants RequiredSamples and WindowSize, along with comments to clarify their purpose.  I’ve kept them as constants since they are intrinsic to the TD Sequential logic and not typically variable. 5 is the minimum needed for setup phases evaluation, and 10 provides enough buffer for lookbacks like perfect setups and countdowns.

I’ve kept them as constants since making them configurable might give the impression they can be tuned arbitrarily, which could lead to misinterpretation of the indicator’s purpose. Happy to revisit if you think exposing them adds anyvalue.

Feel free to suggest !

rathnagiri commented on Indicators/TdSequential.cs:
```diff
@@ -0,0 +1,313 @@
+                    _tdstResistance = _bars.Skip(_bars.Count - MaxSetupCount).Take(MaxSetupCount).Max(b => b.High);
+                    _setupCount = 0;
+
+                    return EncodeState(isPerfect ? TdSequentialPhase.BuySetupPerfect : TdSequentialPhase.BuySetup, 9);
```
> * `9` magic nine again, ideally not hardcoded

Good point about the hardcoded 9. replaced it with MaxSetupCount

rathnagiri replied:
> * HandleBuySetupPhase/HandleSellSetupPhase have quite a bit in common, think they could probably be merged at least some parts of them. Probably same for HandleBuyCountDown/HandleSellCountDown. I think once the variables are merged too it will be much clear the duplication and path for cleanup

Regarding merging the HandleBuy/SellSetupPhase and HandleBuy/SellCountdown methods: I initially explored that direction, but found that while the structure is similar, the control flow diverges in subtle but meaningful ways (e.g., Close < vs Close >, Max(High) vs Min(Low), and distinct phase transitions). Attempting to generalize them increased complexity and made the logic harder to read or follow and maintain.

Since these phases are quite domain-specific, keeping them separate improves readability and makes future enhancements or debugging easier. That said, if more logic ends up being shared over time (especially with variables or phase transitions), I’m happy to revisit and abstract out the common parts where it truly simplifies things.

rathnagiri commented on Indicators/TdSequential.cs:
```diff
@@ -0,0 +1,313 @@
+
+        private bool IsSellSetupPerfect()
+        {
+            if (_bars.Count < MaxSetupCount) return false;
```
Good point ., cleaned up that redundant condition

rathnagiri commented:
Hi @Martin-Molinero 

I believe I have now addressed all the review comments, including the suggestions around method consolidation, hardcoded constants, and the indicator value encoding. I’ve also added clarifications where applicable to explain some of the design decisions.

When you get a chance, please take another look and let me know if anything else needs attention.

Thanks again for your valuable feedback!

Martin-Molinero commented on Indicators/TomDemarkSequential.cs:
```diff
@@ -0,0 +1,353 @@
+        return Default;
+    }
+
+    private decimal InitializeSetupPhase(IBaseDataBar current, IBaseDataBar bar4Ago)
```
From my understanding I think this logic is missing a price flip check, mind double checking this? ref https://practicaltechnicalanalysis.blogspot.com/2013/01/tom-demark-sequential.html

rathnagiri replied:
@Martin-Molinero : I research on this further since the original version in https://demark.com/sequential-indicator/ didn’t include the price flip logic, I wrongly assumed price flip actually the initial 9bar setup. I now added this logic to detect bullish or bearish price flip's before determining the setup phase. That’s also why RequiredSamples was bumped to 6 — we need both the previous bar and the one 4 bars before that to detect a flip.

Martin-Molinero commented on Indicators/TomDemarkSequential.cs:
```diff
@@ -0,0 +1,353 @@
+        var bar9 = window[0];
+
+        return bar8.High >= bar6.High && bar8.High >= bar7.High ||
+            bar9.High >= bar6.High && bar9.High >= bar7.High;
```
Should these be just `>`/`<`  not sure the `=` is valid in the definition? ref https://practicaltechnicalanalysis.blogspot.com/2013/01/tom-demark-sequential.html

Martin-Molinero commented on Indicators/TomDemarkSequential.cs:
```diff
@@ -0,0 +1,353 @@
+
+    private TomDemarkSequentialPhase _currentPhase;
+    
+    private decimal _resistancePrice; // highest high of the 9-bar Tom Demark Sequential buy setup (indicates resistance)
+    private decimal _supportPrice; // lowest low of the 9-bar Tom Demark Sequential sell setup (indicates support)
```
Let's make these a public readonly property, they seems interesting price points to be consumed 👍

Martin-Molinero reviewed (approved):
Thank you! Leaving some last comments

rathnagiri commented on Algorithm/QCAlgorithm.Indicators.cs:
```diff
@@ -2198,6 +2198,22 @@ public TargetDownsideDeviation TDD(Symbol symbol, int period, double minimumAcce
+        /// <param name="resolution">The resolution</param>
+        /// <param name="selector">Selects a value from the BaseData to send into the indicator, if null defaults to casting the input value to a TradeBar</param>
+        /// <returns>The TomDemark Sequential indicator for the requested symbol over the specified period</returns>
+        public TomDemarkSequential TDS(Symbol symbol, Resolution? resolution = null, Func<IBaseData, TradeBar> selector = null)
```
done

rathnagiri commented on Indicators/TdSequential.cs:
```diff
@@ -0,0 +1,313 @@
+
+        private decimal EncodeState(TdSequentialPhase phase, int step)
+        {
+            return (decimal)phase + (step / 100m);
```
I’ve updated ComputeNextValue() to return just the phase enum. Additionally, I’ve exposed two separate public properties for SetupPhaseStepCount and CountdownPhaseStepCount, since step values can overlap (e.g., bar 9 of the setup phase might also be bar 1 of the countdown phase and we need to track them). This distinction helps users accurately track progress using the above properties in each phase without ambiguity.

Martin-Molinero commented on Indicators/TomDemarkSequential.cs:
```diff
@@ -115,7 +130,8 @@ public TomDemarkSequential(string name = "TomDemarkSequential")
     {
-        ResetStepAndPhase();
+        _setupStepCount = 0;
+        _currentPhase = TomDemarkSequentialPhase.None;
```
Missing to reset some of the private fields I believe, like `_countdownStepCount/_supportPrice/_resistancePrice..`

rathnagiri replied:
done

Martin-Molinero commented on Indicators/TomDemarkSequential.cs:
```diff
@@ -87,6 +89,18 @@ public class TomDemarkSequential : WindowIndicator<IBaseDataBar>
+    /// Gets the current Countdown step count in the active TomDemark Sequential (Buy/Sell) Countdown phase
+    /// (e.g., 1 to 13 in Setup), 0 if bar is not in Countdown phase.
+    /// </summary>
+    public int CountdownPhaseStepCount => _countdownStepCount;
```
minor might be simpler/cleaner if the property just had a public get and a private set, no need to define the backing filed manually for these -> will be able to delete a few lines

Martin-Molinero reviewed (commented):
Looks good! just minor bug in reset
